src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -157,6 +157,81 @@ def check_non_empty_list_of_ints_in_range(min_value: int, max_value: int, max_si
     return check_list_of_ints_in_range(min_value, max_value, min_size=1, max_size=max_size, allow_null=allow_null)
 
 
+def separate_endpoint_types(endpoint_dict: dict[int, Any]) -> tuple[list[int], list[int]]:
+    """Returns a tuple containing the list of flat endpoints and a list of tree endpoints"""
+    flat = []
+    tree = []
+    for endpoint_id, endpoint in endpoint_dict.items():
+        if endpoint_id == 0:
+            continue
+        aggregator_id = 0x000e
+        device_types = [d.deviceType for d in endpoint[Clusters.Descriptor][Clusters.Descriptor.Attributes.DeviceTypeList]]
+        if aggregator_id in device_types:
+            flat.append(endpoint_id)
+        else:
+            tree.append(endpoint_id)
+    return (flat, tree)
+
+
+def get_all_children(endpoint_id, endpoint_dict: dict[int, Any]):
+    """Returns all the children (include subchildren) of the given endpoint
+       This assumes we've already checked that there are no cycles, so we can do the dumb things and just trace the tree
+    """
+    children = []
+
+    def add_children(endpoint_id, children):
+        immediate_children = endpoint_dict[endpoint_id][Clusters.Descriptor][Clusters.Descriptor.Attributes.PartsList]
+        if len(immediate_children) == 0:
+            return
+        children.extend(immediate_children)
+        for child in immediate_children:
+            add_children(child, children)
+
+    add_children(endpoint_id, children)
+    return children
+
+
+def find_tree_roots(tree_endpoints: list[int], endpoint_dict: dict[int, Any]) -> set[int]:
+    """Returns a set of all the endpoints in tree_endpoints that are roots for a tree (not include singletons)"""
+    tree_roots = []
```

Comment:
how about setting this as a set directly to start with?

Reply:
yes...that would have been the sensible way to do that.... done.